gtsfm/utils/alignment.py

```diff
@@ -0,0 +1,310 @@
+
+logger = logger_utils.get_logger()
+
+MAX_ALIGNMENT_ERROR = np.finfo(np.float32).max
+MAX_NUM_HYPOTHESIS_FOR_ROBUST_ALIGNMENT: int = 200
+
+
+def align_rotations(aRi_list: List[Optional[Rot3]], bRi_list: List[Optional[Rot3]]) -> List[Rot3]:
+    """Aligns the list of rotations to the reference list by using Karcher mean.
+
+    Args:
+        aRi_list: Reference rotations in frame "a" which are the targets for alignment
+        bRi_list: Input rotations which need to be aligned to frame "a"
+
+    Returns:
+        aRi_list_: Transformed input rotations previously "bRi_list" but now which
+            have the same origin as reference (now living in "a" frame)
+    """
+    aRb_list = [
+        aRi.compose(bRi.inverse()) for aRi, bRi in zip(aRi_list, bRi_list) if aRi is not None and bRi is not None
+    ]
+    if len(aRb_list) > 0:
+        aRb = gtsam.FindKarcherMean(Rot3Vector(aRb_list))
+    else:
+        aRb = Rot3()
+
+    # Apply the coordinate shift to all entries in input.
+    return [aRb.compose(bRi) if bRi is not None else None for bRi in bRi_list]
+
+
+def align_poses_sim3_ignore_missing(
+    aTi_list: List[Optional[Pose3]], bTi_list: List[Optional[Pose3]]
+) -> Tuple[List[Optional[Pose3]], Similarity3]:
+    """Align by similarity transformation, but allow missing estimated poses in the input.
+
+    Note: this is a wrapper for align_poses_sim3() that allows for missing poses/dropped cameras.
+    This is necessary, as align_poses_sim3() requires a valid pose for every input pair.
+
+    We force SIM(3) alignment rather than SE(3) alignment.
+    We assume the two trajectories are of the exact same length.
+
+    Args:
+        aTi_list: reference poses in frame "a" which are the targets for alignment
+        bTi_list: input poses which need to be aligned to frame "a"
+
+    Returns:
+        aTi_list_: transformed input poses previously "bTi_list" but now which
+            have the same origin and scale as reference (now living in "a" frame)
+        aSb: Similarity(3) object that aligns the two pose graphs.
+    """
+    assert len(aTi_list) == len(bTi_list)
+
+    # only choose target poses for which there is a corresponding estimated pose
+    corresponding_aTi_list = []
+    valid_camera_idxs = []
+    valid_bTi_list = []
+    for i, bTi in enumerate(bTi_list):
+        if bTi is not None:
+            valid_camera_idxs.append(i)
+            valid_bTi_list.append(bTi)
+            corresponding_aTi_list.append(aTi_list[i])
+
+    valid_aTi_list_, aSb = align_poses_sim3_robust(aTi_list=corresponding_aTi_list, bTi_list=valid_bTi_list)
+
+    num_cameras = len(aTi_list)
+    # now at valid indices
+    aTi_list_ = [None] * num_cameras
+    for i in range(num_cameras):
+        if i in valid_camera_idxs:
+            aTi_list_[i] = valid_aTi_list_.pop(0)
+
+    return aTi_list_, aSb
+
+
+def align_poses_sim3_exhaustive(aTi_list: List[Pose3], bTi_list: List[Pose3]) -> Tuple[List[Pose3], Similarity3]:
```

Comment:
hmm i thought you limited to 200 hypotheses? exhaustive seems prohibitive for any significant dataset; do we need this fn, or can we just use the variant below named "robust"?

Reply:
I kept the exhaustive just in case. I am also calling this function from robust when the dataset is small enough.

Reply:
Marking as resolved, thanks